### PR TITLE
8277358: Accelerate CRC32-C

### PR DESCRIPTION
**Pull request by asgibbons**
Accelerates CRC32-C by utilizing vpclmulqdq similarly to CRC32.  This change achieves ~4x throughput improvement.

5986.947899319073 MB/s => 24041.05203089616 MB/s
5840.02689336947 MB/s => 24898.781468710356 MB/s

********** Original ***********

```
scottgi@96974-ICX32:~/crc/jdk (asgibbons-crc32c)$ java test/hotspot/jtreg/compiler/intrinsics/zip/TestCRC32C.java 20000000
 offset = 0
msgSize = 512 bytes
  iters = 20000000
-------------------------------------------------------
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
CRC32C.update(byte[]) runtime = 1.710387358 seconds
CRC32C.update(byte[]) throughput = 5986.947899319073 MB/s
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
-------------------------------------------------------
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
CRC32C.update(ByteBuffer) runtime = 1.753416583 seconds
CRC32C.update(ByteBuffer) throughput = 5840.02689336947 MB/s
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
-------------------------------------------------------
```

*********** With my changes: *************

```
scottgi@96974-ICX32:~/crc/jdk (asgibbons-crc32c)$ java test/hotspot/jtreg/compiler/intrinsics/zip/TestCRC32C.java 20000000
 offset = 0
msgSize = 512 bytes
  iters = 20000000
-------------------------------------------------------
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
CRC32C.update(byte[]) runtime = 0.425938099 seconds
CRC32C.update(byte[]) throughput = 24041.05203089616 MB/s
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
-------------------------------------------------------
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
CRC32C.update(ByteBuffer) runtime = 0.411265106 seconds
CRC32C.update(ByteBuffer) throughput = 24898.781468710356 MB/s
CRCs: crc = ae10ee5a, crcReference = ae10ee5a
-------------------------------------------------------
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277358](https://bugs.openjdk.java.net/browse/JDK-8277358): Accelerate CRC32-C

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to 92b4b9fc17a6b53e07017cfad8b5a3e74e8892fe
 * [Eric Caspole](https://openjdk.java.net/census#ecaspole) (@ericcaspole - Committer)

### Contributors
 * Greg Tucker `<greg.b.tucker@intel.com>`
 * Scott Gibbons `<sgibbons@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6595/head:pull/6595` \
`$ git checkout pull/6595`

Update a local copy of the PR: \
`$ git checkout pull/6595` \
`$ git pull https://git.openjdk.java.net/jdk pull/6595/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6595`

View PR using the GUI difftool: \
`$ git pr show -t 6595`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6595.diff">https://git.openjdk.java.net/jdk/pull/6595.diff</a>

</details>

### CODE REVIEW

**Comment by asgibbons** [comment]
Hi, Eric.  I added a microbenchmark for CRC32-C.  I'm waiting for full completion, but it looks like somewhere around 40GB/s throughput on average.  I'll post the results once completed.

**Comment by asgibbons** [comment]
Benchmark results:

```
Benchmark                    (count)  Mode  Cnt  Score    Error  Units
TestCRC32C.testCRC32CUpdate       64  avgt    6  0.021 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate      128  avgt    6  0.031 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate      256  avgt    6  0.023 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate      512  avgt    6  0.026 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate     1024  avgt    6  0.035 ±  0.002  us/op
TestCRC32C.testCRC32CUpdate     2048  avgt    6  0.052 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate     4096  avgt    6  0.092 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate     8192  avgt    6  0.174 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate    16384  avgt    6  0.337 ±  0.001  us/op
TestCRC32C.testCRC32CUpdate    32768  avgt    6  0.663 ±  0.002  us/op
TestCRC32C.testCRC32CUpdate    65536  avgt    6  1.317 ±  0.004  us/op
Finished running test 'micro:java.util.TestCRC32C'

```

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:6588]: a, j,  y and z are only required on the crc32c_ipl_alg2_alt2() path, so should be initialized and saved/restored only there.  
This will also help you to pick a  save on call register like c_rarg3 for table without having to push/pop.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:7218]: We could do this not1(crc) in generate_updateBytesCRC32() thereby remove the need to do the double notl for crc32c.

**Comment by asgibbons** [comment]
Doesn’t this just move the double nots to a different generator?  I’m not comfortable with the cost/benefit of this change.  I don’t want to impact CRC32 for the sake of CRC32C.  I’ll do it if you think it’s worth it.  Please let me know.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:6588]: Done.  Changed to use `j` instead of save/restore of `r14`.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:7218]: Moved the `notl(crc)` calls to `generate_updateBytesCRC32()` as requested.

**Review by vnkozlov — COMMENTED** [review]
Nice work. Let me test it before approval.

**Review by sviswa7 — APPROVED** [review]
The patch looks good to me. Please wait for Vladimir Kozlov's testing and approval.

**Comment by asgibbons** [comment]
Hi, Eric.  Thanks for the suggestions.  I’ve made the changes.

Thanks,
--Scott Gibbons
Software Development Engineer, Runtime Engineering
***@***.***  DEVELOPER SOFTWARE ENGINEERING
Ph: 1-503-456-7756
Cell: 1-469-450-8390
2501 NE Century Blvd
Hillsboro, OR 97124
Intel Corporation | www.intel.com<https://webmail.intel.com/owa/redir.aspx?SURL=WYr7qZDpIv3m1SKFmeHJuzsfCBuGN-jwkBYQUSRR6yrupkscpgzUCGgAdAB0AHAAOgAvAC8AdwB3AHcALgBpAG4AdABlAGwALgBjAG8AbQA.&URL=http%3a%2f%2fwww.intel.com>

**Review by ericcaspole — APPROVED** [review]
The JMH part looks good.
Thanks,
Eric

**Comment by asgibbons** [comment]
@vnkozlov Please let me know if the test passed.

**Review by vnkozlov — APPROVED** [review]
Testing finally passed (at least on x86, aarch64 is still running). I takes long time to get results from several tiers.

**Comment by sviswa7** [comment]
Thanks a lot @vnkozlov.

**Comment by asgibbons** [comment]
Thanks!

**Pull request closed without merging**
